lib/zhuyin.ex

```diff
@@ -0,0 +1,326 @@
+defmodule Zhuyin do
+  @moduledoc """
+  TODO
+  """
+
+  @type t :: %__MODULE__{tone: 0..4, initial: String.t(), final: String.t()}
+
+  @enforce_keys [:final]
+  defstruct tone: 0, initial: "", final: ""
+
+  @typedoc """
+  List of zhuyin syllables mixed with plain strings.
+  """
+  @type zhuyin_list :: [t() | String.t()]
+
+  @zhuyin_tones ["˙", "", "ˊ", "ˇ", "ˋ"]
+
+  def tone_index(nil) do
+    nil
+  end
+  def tone_index(tone_char) do
+    Enum.find_index(@zhuyin_tones, &(&1 == tone_char))
+  end
+
+  @initials %{
+    "ㄅ" => "b",
+    "ㄆ" => "p",
+    "ㄇ" => "m",
+    "ㄈ" => "f",
+    "ㄉ" => "d",
+    "ㄊ" => "t",
+    "ㄋ" => "n",
+    "ㄌ" => "l",
+    "ㄍ" => "g",
+    "ㄎ" => "k",
+    "ㄏ" => "h",
+    "ㄐ" => "j",
+    "ㄑ" => "q",
+    "ㄒ" => "x",
+    "ㄓ" => "zh",
+    "ㄔ" => "ch",
+    "ㄕ" => "sh",
+    "ㄖ" => "r",
+    "ㄗ" => "z",
+    "ㄘ" => "c",
+    "ㄙ" => "s",
+  }
+
+  @standalone_finals %{
+    "ㄧ" => "yi",
+    "ㄨ" => "wu",
+    "ㄩ" => "yu",
+    "ㄧㄚ" => "ya",
+    "ㄨㄚ" => "wa",
+    "ㄧㄥ" => "ying",
+    "ㄧㄤ" => "yang",
+    "ㄧㄝ" => "ye",
+    "ㄨㄛ" => "wo",
+    "ㄨㄥ" => "weng",
+    "ㄨㄤ" => "wang",
+    "ㄧㄠ" => "yaog",
+    "ㄨㄞ" => "waig",
+    "ㄩㄝ" => "yue",
+    "ㄩㄥ" => "yong",
+    "ㄧㄡ" => "you",
+    "ㄨㄟ" => "wei",
+    "ㄧㄢ" => "yan",
+    "ㄨㄢ" => "wan",
+    "ㄩㄢ" => "yuan",
+    "ㄧㄣ" => "yin",
+    "ㄨㄣ" => "wen",
+    "ㄩㄣ" => "yun",
+    # Standalone initials. But parsed as finals for convenience
+    "ㄓ" => "zhi",
+    "ㄔ" => "chi",
+    "ㄕ" => "shi",
+    "ㄖ" => "ri",
+    "ㄗ" => "zi",
+    "ㄘ" => "ci",
+    "ㄙ" => "si",
+    # Standalone finals that are the same as combined
+    "ㄦ" => "er",
+    "ㄢ" => "an",
+  }
+  @finals %{
+    "ㄧ" => "i",
+    "ㄨ" => "u",
+    "ㄩ" => "v",
+    "ㄚ" => "a",
+    "ㄛ" => "o",
+    "ㄜ" => "e",
+    "ㄝ" => "e",
+    "ㄞ" => "ai",
+    "ㄟ" => "ei",
+    "ㄠ" => "ao",
+    "ㄡ" => "ou",
+    "ㄢ" => "an",
+    "ㄣ" => "en",
+    "ㄤ" => "ang",
+    "ㄥ" => "eng",
+    "ㄦ" => "er",
+    "ㄧㄚ" => "ia",
+    "ㄨㄚ" => "ua",
+    "ㄧㄥ" => "ing",
+    "ㄧㄤ" => "iang",
+    "ㄧㄝ" => "ie",
+    "ㄨㄛ" => "uo",
+    "ㄨㄥ" => "ong",
+    "ㄨㄤ" => "uang",
+    "ㄧㄠ" => "iao",
+    "ㄨㄞ" => "uai",
+    "ㄩㄝ" => "ve",
+    "ㄩㄥ" => "iong",
+    "ㄧㄡ" => "iu",
+    "ㄨㄟ" => "ui",
+    "ㄧㄢ" => "ian",
+    "ㄨㄢ" => "uan",
+    "ㄩㄢ" => "van",
+    "ㄧㄣ" => "in",
+    "ㄨㄣ" => "un",
+    "ㄩㄣ" => "vn",
+  }
+
+  @doc"""
+  Convert zhuyin to pinyin
+
+  # Examples
+   iex>  decode_zhuyin("ㄝ")
+   None
+
+   iex>  decode_zhuyin("ㄐㄩ˙")
+   "jv"
+
+   iex>  decode_zhuyin("ㄌㄩˇ")
+   "lv3"
+
+   iex>  decode_zhuyin("ㄓㄨㄢˋ")
+   "zhuan4"
+
+   iex>  decode_zhuyin("ㄓㄠˊ")
+   "zhao2"
+
+   iex>  decode_zhuyin("ㄓˋ")
```

Comment:
I think you have a bit too many doctests here. Write enough doctests to show the concepts that are important to these functions. You can move the other tests to `test/zhuyin_test.exs`.